### PR TITLE
test interaction between drain upon completion & onDrainTimeout

### PR DESCRIPTION
**Pull request by antoniovleonti**
Commit Message: test interaction between drain upon completion & onDrainTimeout
Additional Description: 

adds unit & integration tests that demonstrate it is safe to use max_requests_per_connection with max_connection_duration, idle_timeout, and max_stream_duration.

(idle_timeout is not explicitly tested but triggers the same exact drain behavior as max_connection_duration.)

Risk Level: none
Testing: test-only PR

### CODE REVIEW

File: test/integration/protocol_integration_test.cc [hunk 1]
```diff
@@ -3039,6 +3039,118 @@ TEST_P(DownstreamProtocolIntegrationTest, MaxRequestsPerConnectionReached) {
...
+
+  ASSERT_TRUE(response->waitForEndStream());
+  EXPECT_TRUE(upstream_request_->complete());
+  EXPECT_TRUE(response->complete());
+  EXPECT_EQ(test_server_->counter("http.config_test.downstream_cx_max_requests_reached")->value(),
+            0);
+  EXPECT_EQ(nullptr, response->headers().Connection());
+
+  test_server_->waitForCounterGe("http.config_test.downstream_cx_max_duration_reached", 1);
+  if (downstream_protocol_ != Http::CodecType::HTTP1) {
```
Comment (alyssawilk) [inline test/integration/protocol_integration_test.cc:3069]: can you comment inline why HTTP/1 is not closed?

File: test/integration/protocol_integration_test.cc [hunk 2]
```diff
@@ -3039,6 +3039,118 @@ TEST_P(DownstreamProtocolIntegrationTest, MaxRequestsPerConnectionReached) {
...
+
+  // Sending second request.
+  auto encoder_decoder_2 = codec_client_->startRequest(default_request_headers_);
+  request_encoder_ = &encoder_decoder_2.first;
+  auto response_2 = std::move(encoder_decoder_2.second);
+  codec_client_->sendData(*request_encoder_, 1, true);
+  waitForNextUpstreamRequest();
+
+  // Before sending the response, sleep past the drain timer. Nothing should happen.
+  absl::SleepFor(absl::Seconds(1));
```
Comment (alyssawilk) [inline test/integration/protocol_integration_test.cc:3084]: I can't remember if this test is using synthetic time but I believe timeSystem().advanceTimeWait() is preferred regardless

File: test/integration/protocol_integration_test.cc [hunk 3]
```diff
@@ -3039,6 +3039,118 @@ TEST_P(DownstreamProtocolIntegrationTest, MaxRequestsPerConnectionReached) {
...
+// response!
+TEST_P(DownstreamProtocolIntegrationTest, MaxRequestsPerConnectionVsMaxStreamDuration) {
+  config_helper_.setDownstreamMaxRequestsPerConnection(2);
+  config_helper_.setDownstreamMaxStreamDuration(std::chrono::milliseconds(500));
+
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  // Sending first request and waiting to complete the response.
+  auto encoder_decoder = codec_client_->startRequest(default_request_headers_);
```
Comment (alyssawilk) [inline test/integration/protocol_integration_test.cc:3117]: sendRequestAndWaitForResponse?

File: test/integration/protocol_integration_test.cc [hunk 4]
```diff
@@ -3039,6 +3039,118 @@ TEST_P(DownstreamProtocolIntegrationTest, MaxRequestsPerConnectionReached) {
...
+  EXPECT_TRUE(upstream_request_->complete());
+  EXPECT_TRUE(response->complete());
+  EXPECT_EQ(test_server_->counter("http.config_test.downstream_cx_max_requests_reached")->value(),
+            0);
+
+  // Sending second request and waiting to complete the response.
+  auto encoder_decoder_2 = codec_client_->startRequest(default_request_headers_);
+  request_encoder_ = &encoder_decoder_2.first;
+  auto response_2 = std::move(encoder_decoder_2.second);
+  codec_client_->sendData(*request_encoder_, 1, true);
```
Comment (alyssawilk) [inline test/integration/protocol_integration_test.cc:3134]: auto response2 = codec_client_->makeRequestWithBody(

File: test/integration/protocol_integration_test.cc [hunk 5]
```diff
@@ -3039,6 +3039,118 @@ TEST_P(DownstreamProtocolIntegrationTest, MaxRequestsPerConnectionReached) {
...
+  waitForNextUpstreamRequest();
+  EXPECT_EQ(test_server_->counter("http.config_test.downstream_cx_max_requests_reached")->value(),
+            1);
+
+  // Don't send a response. HCM should sendLocalReply after max stream duration has elapsed.
+  test_server_->waitForCounterGe("http.config_test.downstream_rq_max_duration_reached", 1);
```
Comment (alyssawilk) [inline test/integration/protocol_integration_test.cc:3140]: how long of a wait is that?  if it's more than 1s I'd suggest configuring it down just to make the test less long

File: test/integration/protocol_integration_test.cc [hunk 6]
```diff
@@ -3039,6 +3039,118 @@ TEST_P(DownstreamProtocolIntegrationTest, MaxRequestsPerConnectionReached) {
...
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  // Sending first request and waiting to complete the response.
+  auto encoder_decoder = codec_client_->startRequest(default_request_headers_);
+  request_encoder_ = &encoder_decoder.first;
+  auto response = std::move(encoder_decoder.second);
+  codec_client_->sendData(*request_encoder_, 1, true);
+  waitForNextUpstreamRequest();
+  upstream_request_->encodeHeaders(default_response_headers_, true);
```
Comment (alyssawilk) [inline test/integration/protocol_integration_test.cc:3059]: sendRequestAndWaitForResponse(

**Review by alyssawilk — COMMENTED** [review]
nice testing
/wait

File: test/integration/protocol_integration_test.cc (hunk 5 above)
Comment (antoniovleonti) [reply test/integration/protocol_integration_test.cc:3140]: 500ms

**Comment by antoniovleonti** [comment]
Much better, thanks.

**Review by alyssawilk — COMMENTED** [review]
/wait-any
LGTM modulo that one comment nit

File: test/common/http/conn_manager_impl_test_2.cc [hunk 7]
```diff
@@ -479,6 +479,58 @@ TEST_F(HttpConnectionManagerImplTest, MaxRequests) {
...
+  MockStreamDecoderFilter* filter = new NiceMock<MockStreamDecoderFilter>();
+  EXPECT_CALL(filter_factory_, createFilterChain(_))
+      .WillRepeatedly(Invoke([&](FilterChainManager& manager) -> bool {
+        auto factory = createDecoderFilterFactoryCb(StreamDecoderFilterSharedPtr{filter});
+        manager.applyFilterFactoryCb({}, factory);
+        return true;
+      }));
+
+  startRequest(true);
+  // invoke encodeHeaders -- note we don't end stream.
```
[the remark this answers is not in the document]
Comment (antoniovleonti) [reply test/common/http/conn_manager_impl_test_2.cc]: I misspoke, fixed

**Pull request merged**